Feature/80 投稿機能作成 Postモデル作成

ken-ty opened the pull request:
## 概要
<!-- この PR でやったことの要約や、関連するissueを示してください。 -->

投稿機能作成 ( #88 ) の為、 1投稿である Post モデルを 定義 しました。
#80 は スキップして close していましたが、やはり実装前に必要だったので reopen しました。
issue: #80 

## やったこと
<!-- やったことを箇条書きしてください。多い場合は、PRのスコープを見直してください。 -->

- パッケージ追加
  - Model定義の為にfreezedを導入 [fa3c50e](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/fa3c50e8e8cef53b104bff35b2b86263f10b35f0)
- docs 更新
  - freezedでModelを作成する際のコマンドを追記 [b1e3ce9](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/b1e3ce9ba7a6e7295c0b164b8df4bdc424c08ad2)
 - Postモデルを定義
    - 以下FireStore にあったPostを参考に定義しました [0364857](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/0364857edfc4409ac44e7e114dd7af110e91eae7)
<img width="900" alt="スクリーンショット 2023-11-27 13 31 14" src="https://github.com/Mayumi-Nakabayashi/withTone/assets/38717219/9a61b608-0fc6-42e5-b0c7-b16b7122712b">

- ついでに、PRで自動生成ファイルの差分を目立たなくする設定を追加した。 [8707dd1](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/8707dd1378ef0fb6b4078a546a1c96bc98cfd163)

<img width="900" alt="スクリーンショット 2023-11-27 13 40 37" src="https://github.com/Mayumi-Nakabayashi/withTone/assets/38717219/3d4f8cd4-0faa-46e8-b01f-e6bad9195fca">

---

MTG でメンバーと会話した後、以下の調整をしました。

- [docs: Post の説明を更新](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/f6cf980c94e073be21c1c7bbb466555f99f3384e)
[f6cf980](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/f6cf980c94e073be21c1c7bbb466555f99f3384e)
- [build: firestoreのパッケージをインストール](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/362b46ccf14fbd7a0d136620226a8a614ca3bffe)
[362b46c](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/362b46ccf14fbd7a0d136620226a8a614ca3bffe)
- [feat: 定義したPostで投稿機能を仮作成](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/443b7d2f3cda5db621171eb04e1e0bac03db1411)
[443b7d2](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/443b7d2f3cda5db621171eb04e1e0bac03db1411)
- [feat: いいねはサブコレクションにする](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/08a0ffc85458ae0921d88553cf951ca692d3ecd2)


## 相談

Post モデルのフィールドに過不足がないか確認したいです。
どこかで meet して話しながら議論できると。

- 作成者や更新日時は不要なのか
- uid について、認識合わせしたい。
- movieUrl に画像を入れているが、画像・動画どちらも入るのか
- タグはStringか ( タグをテーブル管理するなら、タグIDが入る？ ）

→ 会話して、調整した。

データ型を作成し、FireStore に仮でアップロードまでやった。

## キャプチャ

仮作成した投稿メソッドで送信されたPostデータ

<img width="900" alt="スクリーンショット 2023-12-01 4 05 28" src="https://github.com/Mayumi-Nakabayashi/withTone/assets/38717219/166be22e-045e-48a2-9099-8590098538a3">

ken-ty committed:
build: Model定義の為にfreezedを導入

ken-ty committed:
docs: freezedでModelを作成する際のコマンドを追記

ken-ty committed:
feat: 投稿データモデルをPostとして仮作成

FireStoreに誰かが定義した内容ままで作成.作成者や更新日時など不足してそう.

ken-ty committed:
chore: 自動生成ファイルをPRで目立たなくする設定を追加

ken-ty commented on lib/models/post.dart:
```diff
@@ -0,0 +1,39 @@
+    required String id,
+
+    // uid は、投稿したユーザーのID??
+    required String uid,
```
firebase の ユーザーの ID 
ログインした時に取得できる。
uid = userId

ken-ty commented on lib/models/post.dart:
```diff
@@ -0,0 +1,39 @@
+    required String uid,
+
+    /// タイトル
+    required String title,
```
"" が入る可能性あり

ken-ty commented on lib/models/post.dart:
```diff
@@ -0,0 +1,39 @@
+    /// 動画のURL
+    ///
+    /// 画像が入る可能性もある? 分けた方がよさそう
+    required String movieUrl,
```
画像だったら1枚、
動画も1、
画像+動画はなし

ken-ty replied:
画像が複数の可能性あり

ken-ty replied:
一旦、動画1だけ...

ken-ty replied:
画像はPh2

ken-ty commented on lib/models/post.dart:
```diff
@@ -0,0 +1,39 @@
+    required String movieUrl,
+
+    /// お気に入りしたユーザー
+    required List<String> favoriteUsers,
```
uid のリスト

ken-ty replied:
post の サブコレクションで作る

ken-ty replied:
get post/${id}/favorites
とかでいいね数とる

ken-ty replied:
一旦このまま作っちゃう、Ph2 とかで分けるか検討する？

ken-ty replied:
サブコレクションにするの大変じゃなかったので、やりました。

ken-ty commented on lib/models/post.dart:
```diff
@@ -0,0 +1,39 @@
+    required List<String> favoriteUsers,
+
+    /// 紐づけられたタグ一覧
+    required List<String> tags,
```
文字列で。

ken-ty commented on lib/models/post.dart:
```diff
@@ -0,0 +1,39 @@
+    required List<String> tags,
+
+    /// 作成日時
+    required DateTime createdAt,
```
更新日時 ... 入れる？、入れてもいい。。
いずれ使えるかも、一旦入れる。

ken-ty replied:
コメント更新されても、Postの更新はされない。
Post の本文、タイトル、、、などが更新されたらupdatedAt を変更する。
コメントやいいねが更新されても更新しない。

sodateya replied:
firebaseと時間を送ったりする際はTimeStamp型でやりとりする形になります。
firebaseからとってきたTimeStampをFlutterで使えるようにDateTimeに変換する処理が必要になります。

良さげな記事ありましたので、こちら参考にしてみてください！
https://qiita.com/GleamingCake/items/4ebdc27a4dc3cbd07458

ken-ty replied:
ありがとうございます！対応しました。
[feat: Firestoreで日時はTimestampFieldなのでConerterを用意](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/34a920102d372a7c9866c63c08e4b123689e5bab)
[34a9201](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/34a920102d372a7c9866c63c08e4b123689e5bab)

ken-ty committed:
docs: Post の説明を更新

ken-ty committed:
Merge branch 'develop' of ssh://github.com/Mayumi-Nakabayashi/withTone into feature/88-create-post-func

ken-ty committed:
build: firestoreのパッケージをインストール

ken-ty committed:
feat: 定義したPostで投稿機能を仮作成

ken-ty committed:
feat: いいねはサブコレクションにする

sodateya commented on lib/views/pages/upload_commentq/upload_commentq_page.dart:
```diff
@@ -47,6 +50,42 @@ class _UploadCommentqPageState extends State<UploadCommentqPage> {
+                // 仮投稿作成
+                // documentId が欲しいので、一度 add してから update する
+                FirebaseFirestore.instance.collection('posts').add({}).then(
+                  (docRef) {
```
addしてからupdateするとupdate分のコストが余分にかかってしますので、一度doc()までインスタンスを生成してしまってdocIDを取得して、addではなくsetするような形が良いかもしれません。

```
final doc =  FirebaseFirestore.instance.collection('posts').doc();
final docId = doc.id;

/////省略/////

doc.set({post.toJson});

```
みたいな形にすると良いかと思います。

現在仮実装ということですが本実装では、firebaseへの投稿、取得をwithConverterを使うと何かと便利ですので、やってみてください！

参考
firebaseのadd,setなど　https://zenn.dev/ymktmk/articles/be9a8b209daf3a
withConverter　https://zenn.dev/tsuruo/articles/23894990188653

ken-ty replied:
add + update を set に変えました！

[perf: add して update せず set で 1回だけアクセスする](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/d65f68bb834b62f47771199441ecd4324219afe8)
[d65f68b](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/d65f68bb834b62f47771199441ecd4324219afe8)

ken-ty committed:
chore(analyzer): freezed使用時に誤検知でwarningになるlintルールを無視する

https://pub.dev/documentation/freezed/latest/#disabling-invalid_annotation_target-warning-and-warning-in-generates-files
に記載されている

ken-ty commented:
追加：

warning でたので。公式docsにも無効化してくれと書いてあるlintルールを無効化した。

<img width="563" alt="スクリーンショット 2023-12-03 14 53 27" src="https://github.com/Mayumi-Nakabayashi/withTone/assets/38717219/36d5afee-841c-4276-9104-37d4e5d8d3cf">


[chore(analyzer): freezed使用時に誤検知でwarningになるlintルールを無視する](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/9616997195a13d114dfa1eca742898010c6154de) 
[9616997](https://github.com/Mayumi-Nakabayashi/withTone/pull/149/commits/9616997195a13d114dfa1eca742898010c6154de)
https://pub.dev/documentation/freezed/latest/#disabling-invalid_annotation_target-warning-and-warning-in-generates-files
に記載されている

ken-ty committed:
feat: Firestoreで日時はTimestampFieldなのでConerterを用意

ken-ty committed:
perf: add して update せず set で 1回だけアクセスする

sodateya commented on lib/views/pages/upload_commentq/upload_commentq_page.dart:
```diff
@@ -51,41 +51,31 @@ class _UploadCommentqPageState extends State<UploadCommentqPage> {
+                  movieUrl: 'https://www.youtube.com/watch?v=8HqQ3XgxBhY',
+                  tags: ['#バイオリン', '#初心者', '#質問歓迎'],
+                  // これは自動で入れたい
+                  createdAt: DateTime.now(), // これだとサーバー時刻じゃない
```
ユーザの端末の時間がずれていたら変な時間になってしまうので、 できたらサーバータイムの方が良いと思います！
cloud_firestoreのパッケージの中ににFieldValue.serverTimestamp()というものがありますので、こちらで実装すると良いかもしれません！

ken-ty replied:
こちら修正しました！ @sodateya さんが作ってくれた Converter を cherry-pick して置き換えました！

sodateya committed:
feat:DataTime=>TimeStamp作成(ken-tyさんが作成済。不要になったら削除)

ken-ty committed:
fix: クライアントの時刻ではなく、サーバーの時刻を反映するよう修正
